src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -188,6 +189,132 @@ public partial class JsonTypeInfo<T>
+
+                try
+                {
+                    SerializeHandler(writer, rootValue!);
+                    writer.Flush();
+                }
+                finally
+                {
+                    // Record the serialization size in both successful and failed operations,
+                    // since we want to immediately opt out of the fast path if it exceeds the threshold.
+                    OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
+
+                    Utf8JsonWriterCache.ReturnWriter(writer);
+                }
+
+                await utf8Json.FlushAsync(cancellationToken).ConfigureAwait(false);
+            }
+            else if (
+#if NETCOREAPP
+                !typeof(T).IsValueType &&
+#endif
+                Converter.CanBePolymorphic &&
+                rootValue is not null &&
+                Options.TryGetPolymorphicTypeInfoForRootType(rootValue, out JsonTypeInfo? derivedTypeInfo))
+            {
+                Debug.Assert(typeof(T) == typeof(object));
+                await derivedTypeInfo.SerializeAsObjectAsync(utf8Json, rootValue, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                bool isFinalBlock;
+                WriteStack state = default;
+                state.Initialize(this,
+                    rootValueBoxed,
+                    supportContinuation: true,
+                    supportAsync: true);
+
+                state.CancellationToken = cancellationToken;
+
+                using var writer = new Utf8JsonWriter(utf8Json, Options.GetWriterOptions());
+
+                try
+                {
+                    do
+                    {
+                        // Arbitrary threshold, default Pipe buffers are 4k so this allows 4 buffers to be created and filled (mostly) before flushing.
+                        state.FlushThreshold = (int)(16384 * JsonSerializer.FlushThreshold);
+
+                        try
+                        {
+                            isFinalBlock = EffectiveConverter.WriteCore(writer, rootValue, Options, ref state);
+                            writer.Flush();
+
+                            if (state.SuppressFlush)
+                            {
+                                Debug.Assert(!isFinalBlock);
+                                Debug.Assert(state.PendingTask is not null);
+                                state.SuppressFlush = false;
+                            }
+                            else
+                            {
+                                await utf8Json.FlushAsync(cancellationToken).ConfigureAwait(false);
+                            }
+                        }
+                        finally
+                        {
+                            // Await any pending resumable converter tasks (currently these can only be IAsyncEnumerator.MoveNextAsync() tasks).
```

Comment:
This method is copying over a lot of delicate code from the `Stream` implementation, so the two would have to always be kept in sync. Is there any chance we could apply a bit of DRY here (perhaps by shimming both `PooledByBufferWriter` and `Stream` behind a single type that handles flushing concerns?)

Reply:
It shouldn't in principle -- the cost of resuming the serializer after every flush operation is the dominant factor by a substantial margin. It would be nice if we this could be done without incurring additional allocations though.

Reply:
I care about in practice 😄

Reply:
Here is an attempt: https://github.com/BrennanConroy/runtime/commit/0cc2645f000bd716e55da0c57b7c37b2479955fd

It's heavily influenced by similar patterns e.g. https://github.com/dotnet/runtime/blob/5b4e770daa190ce69f402c1cf92f388b07d3e144/src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs#L454

Haven't convinced myself that it's actually nicer though.

Reply:
I think it's a step in the right direction. I left some feedback on the commit for potential simplifications.

Reply:
Updated change https://github.com/dotnet/runtime/compare/9448b38db216fb04126836fb7319d5c4e307667f...BrennanConroy:runtime:da6d4496f2dd23129f49e9f0c3f1ec19df18a692

Reply:
LGTM. Guessing it hasn't been merged to the PR branch yet?